nemo/collections/tts/modules/radtts.py

```diff
@@ -0,0 +1,870 @@
+            if self.ap_use_voiced_embeddings:
+                self.v_embeddings = torch.nn.Embedding(4, n_text_dim)
+
+        if 'apm' in include_modules:
+            f0_model_config['hparams']['n_speaker_dim'] = n_speaker_dim
+            energy_model_config['hparams']['n_speaker_dim'] = n_speaker_dim
+            if self.use_first_order_features:
+                f0_model_config['hparams']['n_in_dim'] = 2
+                energy_model_config['hparams']['n_in_dim'] = 2
+                if (
+                    'spline_flow_params' in f0_model_config['hparams']
+                    and f0_model_config['hparams']['spline_flow_params'] is not None
+                ):
+                    f0_model_config['hparams']['spline_flow_params']['n_in_channels'] = 2
+                if (
+                    'spline_flow_params' in energy_model_config['hparams']
+                    and energy_model_config['hparams']['spline_flow_params'] is not None
+                ):
+                    energy_model_config['hparams']['spline_flow_params']['n_in_channels'] = 2
+            else:
+                if (
+                    'spline_flow_params' in f0_model_config['hparams']
+                    and f0_model_config['hparams']['spline_flow_params'] is not None
+                ):
+                    f0_model_config['hparams']['spline_flow_params']['n_in_channels'] = f0_model_config['hparams'][
+                        'n_in_dim'
+                    ]
+                if (
+                    'spline_flow_params' in energy_model_config['hparams']
+                    and energy_model_config['hparams']['spline_flow_params'] is not None
+                ):
+                    energy_model_config['hparams']['spline_flow_params']['n_in_channels'] = energy_model_config[
+                        'hparams'
+                    ]['n_in_dim']
+
+            self.f0_pred_module = get_attribute_prediction_model(f0_model_config)
+            self.energy_pred_module = get_attribute_prediction_model(energy_model_config)
+
+    def encode_speaker(self, spk_ids):
+        spk_ids = spk_ids * 0 if self.dummy_speaker_embedding else spk_ids
+        spk_vecs = self.speaker_embedding(spk_ids)
+        return spk_vecs
+
+    def encode_text(self, text, in_lens):
+        # text_embeddings: b x len_text x n_text_dim
+        text_embeddings = self.embedding(text).transpose(1, 2)
+        # text_enc: b x n_text_dim x encoder_dim (512)
+        if in_lens is None:
+            text_enc = self.encoder.infer(text_embeddings).transpose(1, 2)
+        else:
+            text_enc = self.encoder(text_embeddings, in_lens).transpose(1, 2)
+
+        return text_enc, text_embeddings
+
+    def preprocess_context(self, context, speaker_vecs, out_lens=None, f0=None, energy_avg=None):
+
+        if self.n_group_size > 1:
+            context = self.unfold(context.unsqueeze(-1))
+            # (todo): fix unfolding zero-padded values
```

Comment:
How urgent is this TODO? Does this mess with batch training?

Reply:
not doing anything. removed comment